### ISSUE TITLE
scripts/create_allowlist.sh doesn't produce a boot_aggregate item and IMA attestation requires it

### ISSUE THREAD

**Issue by dnoliver** [keylime/keylime#546]
Is your an issue a feature request? If so, please raise it as an [enhancement](https://github.com/keylime/enhancements)

I believe it is an issue with the current implementation.

# Environment
 * OS / version: Fedora Server 33
 * Processor architecture: x86_64
 * TPM Manufacturer: Microsoft HyperV vTPM
 * Keylime version: keylime-5.8.0-1.fc33.noarch

# Description 

`scripts/create_allowlist.sh` does not add a boot_aggregate item, nor have an option to add the current one.
After generating the allowlist and using `keylime_tenant` to set up IMA attestation, I get the following message:

```shell
2021-02-12 16:03:32.281 - keylime.ima - WARNING - No boot_aggregate value found in allowlist, adding an empty one
```

Then, attestation fails because it compares against 0.

To workaround this, add a "boot_aggregate" item with the corresponding digest. To add the current boot_aggregate for the current system:

```shell
grep boot_aggregate /sys/kernel/security/ima/ascii_runtime_measurements | awk '{ sub("sha256:","",$4); print $4 "  " $5 }' >> allowlist.txt
```

## Expected behavior vs. actual behavior 

Expected behaviour is one of the following:

1. You should not need to add a boot_aggregate in allowlist. The expected boot_aggregate should be taken from the TPM Quote, by concatenating and digesting the reported PCR 0 to 7 values. The IMA Measurement Log boot_aggregate should be compared with the recontructed boot_aggregate from the TPM Quote. If that matches, then we can start trusting whatever is in the IMA Measurement Log
2. The tool should warn about the missing boot_aggregate item, or ask to include the current one (it is taking the current digest of the binaries, so getting the current boot_aggregate make sense here as well)
3. Reconstruct the boot_aggregate from the current PCR values an add it to the allowlist?

## Steps to reproduce problem
1.  Run `scripts/create_allowlist.sh allowlist.txt sha256sum`
2.  Use `allowlist.txt` to add a host with the `keylime_tenant` command
3.  IMA Attestation will fail 

# Relevant logs

```shell
2021-02-12 16:03:32.281 - keylime.ima - WARNING - No boot_aggregate value found in allowlist, adding an empty one
```

Attach any relevant log files that can help to debug your issue.

**Comment by dnoliver** [comment]
I just noticed that enabling IMA attestation doesn't make Keylime to quote PCR 0 to 7. 
As shown in the log below, the quote is being done over `sha256:15,16,22+sha1:10`.
I think it should be mandatory to quote PCR 0 to 7 when IMA is enabled, so you can reconstruct the `boot_aggregate` from the quote in the verifier, and "extend the Hardware Root of Trust to the userspace components". 

```shell
2021-02-12 17:37:14.121 - keylime.tpm2 - INFO - ['tpm2_quote', '-c', '/var/lib/keylime/secure/tmpday1o0pw', '-l', 'sha256:15,16,22+sha1:10', '-q', '57426f3378385035504930706c525048434d7662', '-m', '/tmp/tmpeq29tc4p', '-s', '/tmp/tmptmc8adxd', '-o', '/tmp/tmph1x5gwjn', '-g', 'sha256', '-p', '8kTx1QOykcCss9W8yvYP']
2021-02-12 17:37:14.211 - keylime.cloudagent - INFO - GET integrity quote returning 200 response.
```

**Comment by nabilschear** [comment]
hi, so the boot aggregate is a roll-up of those lower PCRs, so you can use that.  It's a bit complicated to compute, but there is some code to do it.  http://linux-ima.sourceforge.net/linux-ima-measurements.html  You're right though that create_whitelist.sh doesn't check this.  You can add it by hand to your whitelist using the above kernel test code.

**Comment by lukehinds** [comment]
sounds like a good change to me, or you up for amending the create_allowlist @dnoliver ?

**Comment by dnoliver** [comment]
I think that, to solve the immediate problem of having a demo that just works, adding some commands in `scripts/create_allowlist.sh` to append the current boot_aggregate (taken from `/sys/kernel/security/ima/ascii_runtime_measurement` or computed from the output of `tpm2_pcrread`) to the allowlist output works. I think I can do that @lukehinds :) 

But for the future, I really think that keylime should quote PCR 0 to 7 when the `--allowlist` is used, compute the boot_aggregate from the quote, and use that value for comparison with the boot_aggregate reported in the IMA measurement log. But this is a much bigger change. 

I was also going to say something about the SHA1 bank being used by default, arguing that Fedora now uses SHA256 for everything related to IMA by default, but I noticed that still defaults to SHA1 in the PCR bank selection :(. I filed a bug for this in https://bugzilla.redhat.com/show_bug.cgi?id=1928286

**Comment by lukehinds** [comment]
The sha1 is interesting @dnoliver , at first it sounds weak, but then you factor in that IMA uses an 'extend operation' to compute a long chain of measurements (all being concatenated and then hashed again), so this negates the risks of collisions you get with a single sha1, so sha1 in this context is not collision prone. 

@maugustosilva / @galmasi , I seem to remember one of you guys saying you know of a boot aggregate compute tool?

**Comment by dnoliver** [comment]
`tsseventextend` computes the boot aggregate

```shell
[test@fedora-server keylime]$ dnf provides tsseventextend
Last metadata expiration check: 29 days, 6:42:15 ago on Thu 14 Jan 2021 12:05:58 PM -03.
tss2-1331-6.fc33.i686 : IBM's TCG Software Stack (TSS) for TPM 2.0 and related utilities
Repo        : fedora
Matched from:
Filename    : /usr/bin/tsseventextend

[test@fedora-server keylime]$ sudo tsseventextend -sim -if /sys/kernel/security/tpm0/binary_bios_measurements

algorithmId TPM_ALG_SHA1
 PCR 00: ff 4c b8 f5 c1 a3 ae 2b e6 1a 6e 20 01 35 dc 8f a3 44 65 06
 PCR 01: 86 a5 3e 51 ce d3 51 37 32 84 cc a4 0d c1 ab a8 00 02 fa 4d
 PCR 02: b2 a8 3b 0e bf 2f 83 74 29 9a 5b 2b df c3 1e a9 55 ad 72 36
 PCR 03: b2 a8 3b 0e bf 2f 83 74 29 9a 5b 2b df c3 1e a9 55 ad 72 36
 PCR 04: 8a 75 e3 6b be 46 ab 40 e3 74 fc e9 fb d9 67 1d a1 aa 47 ea
 PCR 05: c0 8f 2d ff 0b 54 a1 99 c5 f6 8b 4b 33 10 98 5a 5a a1 f3 e7
 PCR 06: b2 a8 3b 0e bf 2f 83 74 29 9a 5b 2b df c3 1e a9 55 ad 72 36
 PCR 07: c9 f3 62 04 cc fc fb 83 fe 17 6d d8 14 26 9c 0c f4 4d 4f 92
 PCR 08: 69 48 d6 e9 de 47 eb b9 ce cf 1c 1a 05 9c cf d1 67 5d ad 7f
 PCR 09: d1 26 d6 fa 0b eb 33 f5 3b 7f 20 e2 5c fa 5c 2f ea 08 a2 51
 PCR 10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 11: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 12: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 13: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 14: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 15: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 16: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 17: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 18: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 19: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 21: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 22: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 23: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

boot aggregate: 31 d6 0e 68 06 55 27 d6 35 d8 b0 88 d3 d5 e2 05 22 c7 80 d5

algorithmId TPM_ALG_SHA256
 PCR 00: f4 81 e4 1a 47 24 5d 35 b5 38 b4 6c 45 29 13 b5 91 f5 36 a6 a4 d0 07 42 c4 32 f7 b7 eb e9 b5 7e
 PCR 01: f8 5e 9c 00 11 a0 d2 a0 83 e4 c8 a7 91 37 37 1e fc d9 24 51 5d 10 60 f0 7e ed 06 3a 51 1a 6a f3
 PCR 02: 3d 45 8c fe 55 cc 03 ea 1f 44 3f 15 62 be ec 8d f5 1c 75 e1 4a 9f cf 9a 72 34 a1 3f 19 8e 79 69
 PCR 03: 3d 45 8c fe 55 cc 03 ea 1f 44 3f 15 62 be ec 8d f5 1c 75 e1 4a 9f cf 9a 72 34 a1 3f 19 8e 79 69
 PCR 04: 5d e7 0e 51 6e 40 b0 b2 d6 40 4c a2 27 2e ad 9e 9e 3a ae df bf cc 98 a0 b6 a8 85 5f 8c 64 12 14
 PCR 05: cb 72 41 71 b1 68 86 62 a0 fc 01 9e 83 98 22 6a 30 05 08 0b 7b 07 9c 95 b4 96 cf 58 9c ad 0b d4
 PCR 06: 3d 45 8c fe 55 cc 03 ea 1f 44 3f 15 62 be ec 8d f5 1c 75 e1 4a 9f cf 9a 72 34 a1 3f 19 8e 79 69
 PCR 07: 11 96 1b 5c 0a 9f 9c f7 26 dc 96 08 b5 a4 2a 17 d3 4d 1d d5 1a e7 07 ed 12 68 64 cd 80 b1 75 53
 PCR 08: b3 18 e3 3d 0c c0 64 ba 73 5b 5d 52 9e e8 b7 17 f2 9a 38 a3 b8 cb 0d a7 cb 30 a1 15 50 2f 58 23
 PCR 09: 94 e9 32 c5 16 fa 72 dc 4c 7a 80 13 12 94 84 87 0e 95 11 78 5a 86 3e 3b 15 65 23 51 9e 32 70 33
 PCR 10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 11: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 12: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 13: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 14: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 15: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 16: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 17: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 18: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 19: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 21: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 22: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 PCR 23: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

boot aggregate: 25 69 62 34 0b 9c 6b 8f 0b 29 0c 3b 63 2d ed ec 8d ea e7 98 8b 7f c7 38 56 d4 06 54 ac e6 ae b8
```

This repo also provides a tool for computing it: https://github.com/initlove/attest-tools. 

The Linux Test Project also have code a test for it: https://github.com/linux-test-project/ltp/blob/master/testcases/kernel/security/integrity/ima/src/ima_boot_aggregate.c

**Comment by maugustosilva** [comment]
@dnoliver @lukehinds Two quick comments:
1) Yes, you are correct, the scripts we have currently on the repository do not add the boot aggregate as the first item on the `allowlist`, and the solution that you pointed out (grabbing from `ascii_runtime_measurement)` is the correct one. We have been using internally a significantly improved version of the script, and I fully intend to open a PR on this.
2) The computation of the value of PCRs is of utmost importance, and we are doing a non-trivial amount of work to get that incorporated into Keylime as quickly as possible. Please take a look https://github.com/keylime/keylime/issues/491 and you will see that we are including (and validating) PCRs 0-9 (and 11-14) independently of IMA (using "TPM pre-boot event logs"). We have one PR already merged (#517), one waiting to be merged (#534), with two more ready to go.

**Comment by galmasi** [comment]
I believe `tpm2_eventlog` in the Intel TPM2 tool chain will do things similar to tsseventextend. We are using it, in any case, as part of measured boot -- which the boot aggregate represents, in the PRs @maugustosilva is referring to.
